File: old/sorting/merge_sort/kth_element_of_merged_halves/kth_element_of_merged_halves.py
def kth_element_of_merged_halves(nums, mid, k):
    left_arr = nums[:mid]
    right_arr = nums[mid:]
    left_ptr = 0
    right_ptr = 0

    count = 0
    while left_ptr < len(left_arr) and right_ptr < len(right_arr):
        if left_arr[left_ptr] < right_arr[right_ptr]:
            val = left_arr[left_ptr]
            left_ptr += 1
        else:
            val = right_arr[right_ptr]
            right_ptr += 1

        if count == k:
            return val

        count += 1

    while left_ptr < len(left_arr):
        val = left_arr[left_ptr]
        if count == k:
            return val
        count += 1
        left_ptr += 1

    while right_ptr < len(right_arr):
        val = right_arr[right_ptr]
        if count == k:
            return val
        count += 1
        right_ptr += 1

    return None


def main():
    output = kth_element_of_merged_halves([1, 4, 7, 2, 3, 9], 3, 4)
    print(output)


main()

File: old/sorting/merge_sort/kth_element_of_merged_halves/kth_element_of_merged_halves.cpp
#include <iostream>
#include <vector>

using namespace std;

int kth_element_of_merged_halves(vector<int> arr, int mid, int k) {
    vector<int> leftArr(arr.begin(), arr.begin() + mid);
    vector<int> rightArr(arr.begin() + mid, arr.end());
    vector<int> sortedArr;
    int leftPtr = 0;
    int rightPtr = 0;

    while (leftPtr < leftArr.size() && rightPtr < rightArr.size()) {

        if (leftArr[leftPtr] < rightArr[rightPtr]) {
            sortedArr.push_back(leftArr[leftPtr]);
            leftPtr += 1;
        } else {
            sortedArr.push_back(rightArr[rightPtr]);
            rightPtr += 1;
        }
    }


    while (leftPtr < leftArr.size()) {
        sortedArr.push_back(leftArr[leftPtr]);
        leftPtr += 1;
    }


    while (rightPtr < rightArr.size()) {
        sortedArr.push_back(rightArr[rightPtr]);
        rightPtr += 1;
    }

    for (int i = 0; i < sortedArr.size(); i++) {
        if (i == k - 1) {
            return sortedArr[i];
        } 
    }

    return -1;
}

int main() {
    cout << kth_element_of_merged_halves({1, 3, 4, 2, 5, 6}, 3, 4);
    return 0;
}
